File: app/src/main/java/com/hussain/podcastapp/model/Entry.java
package com.hussain.podcastapp.model;

import android.arch.persistence.room.Embedded;
import android.arch.persistence.room.Entity;
import android.arch.persistence.room.Ignore;
import android.arch.persistence.room.PrimaryKey;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;

import com.google.firebase.database.Exclude;
import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;


@Entity(tableName = "entry")
public class Entry implements Parcelable {

    @Exclude
    @SuppressWarnings("unused")
    public static final Parcelable.Creator<Entry> CREATOR = new Parcelable.Creator<Entry>() {
        @Override
        public Entry createFromParcel(Parcel in) {
            return new Entry(in);
        }

        @Override
        public Entry[] newArray(int size) {
            return new Entry[size];
        }
    };
    @NonNull
    @PrimaryKey
    @Embedded
    @SerializedName("id")
    public FeedID feedId;
    @Embedded
    @SerializedName("title")
    public Title EntryTitle;
    @SerializedName("im:image")
    public List<PodcastImage> image;
    @Embedded
    @SerializedName("summary")
    public Summary summary;

    public Entry() {
    }

    public FeedID getFeedId() {
        return feedId;
    }

    public void setFeedId(FeedID feedId) {
        this.feedId = feedId;
    }

    @Ignore
    protected Entry(Parcel in) {
        feedId = (FeedID) in.readValue(FeedID.class.getClassLoader());
        EntryTitle = (Title) in.readValue(Title.class.getClassLoader());
        if (in.readByte() == 0x01) {
            image = new ArrayList<>();
            in.readList(image, PodcastImage.class.getClassLoader());
        } else {
            image = null;
        }
        summary = (Summary) in.readValue(Summary.class.getClassLoader());
    }

    @Exclude
    public Title getEntryTitle() {
        return EntryTitle;
    }

    @Exclude
    public void setEntryTitle(Title entryTitle) {
        this.EntryTitle = entryTitle;
    }

    public List<PodcastImage> getImage() {
        return image;
    }

    public Summary getSummary() {
        return summary;
    }

    public void setSummary(Summary summary) {
        this.summary = summary;
    }

    public void setImage(List<PodcastImage> image) {
        this.image = image;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Exclude
    public Map<String, Object> toMap() {
        HashMap<String, Object> result = new HashMap<>();
        result.put("feedId", feedId);
        result.put("Title", EntryTitle);
        result.put("image", image);
        result.put("summary", summary);
        return result;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeValue(feedId);
        dest.writeValue(EntryTitle);
        if (image == null) {
            dest.writeByte((byte) (0x00));
        } else {
            dest.writeByte((byte) (0x01));
            dest.writeList(image);
        }
        dest.writeValue(summary);
    }

    public static class Title implements Parcelable {

        @Exclude
        @Ignore
        @SuppressWarnings("unused")
        public final Parcelable.Creator<Title> CREATOR = new Parcelable.Creator<Title>() {
            @Override
            public Title createFromParcel(Parcel in) {
                return new Title(in);
            }

            @Override
            public Title[] newArray(int size) {
                return new Title[size];
            }
        };

        public Title() {
        }

        @Ignore
        Title(Parcel in) {
            labelTitle = in.readString();
        }

        @SerializedName("label")
        public String labelTitle;

        public String getLabelTitle() {
            return labelTitle;
        }

        public void setLabelTitle(String labelTitle) {
            this.labelTitle = labelTitle;
        }

        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            dest.writeString(labelTitle);
        }
    }

    public static class Summary implements Parcelable {

        @Exclude
        @Ignore
        @SuppressWarnings("unused")
        public final Parcelable.Creator<Summary> CREATOR = new Parcelable.Creator<Summary>() {
            @Override
            public Summary createFromParcel(Parcel in) {
                return new Summary(in);
            }

            @Override
            public Summary[] newArray(int size) {
                return new Summary[size];
            }
        };
        public String label;

        public String getLabel() {
            return label;
        }

        public void setLabel(String label) {
            this.label = label;
        }

        public Summary() {
        }

        @Ignore
        protected Summary(Parcel in) {
            label = in.readString();
        }

        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            dest.writeString(label);
        }
    }

    public static class FeedID implements Parcelable {

        @Exclude
        @Ignore
        @SuppressWarnings("unused")
        public final Parcelable.Creator<FeedID> CREATOR = new Parcelable.Creator<FeedID>() {
            @Override
            public FeedID createFromParcel(Parcel in) {
                return new FeedID(in);
            }

            @Override
            public FeedID[] newArray(int size) {
                return new FeedID[size];
            }
        };
        @Embedded
        @NonNull
        public Attributes attributes;

        public Attributes getAttributes() {
            return attributes;
        }

        public void setAttributes(Attributes attributes) {
            this.attributes = attributes;
        }

        public FeedID() {
        }

        @Ignore
        FeedID(Parcel in) {
            attributes = (Attributes) in.readValue(Attributes.class.getClassLoader());
        }

        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            dest.writeValue(attributes);
        }
    }

    public static class Attributes implements Parcelable {

        @Exclude
        @Ignore
        @SuppressWarnings("unused")
        public final Parcelable.Creator<Attributes> CREATOR = new Parcelable.Creator<Attributes>() {
            @Override
            public Attributes createFromParcel(Parcel in) {
                return new Attributes(in);
            }

            @Override
            public Attributes[] newArray(int size) {
                return new Attributes[size];
            }
        };
        @NonNull
        @SerializedName("im:id")
        public String id;

        public Attributes() {
        }

        @Ignore
        protected Attributes(Parcel in) {
            id = in.readString();
        }

        public String getId() {
            return id;
        }

        public void setId(String id) {
            this.id = id;
        }

        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            dest.writeString(id);
        }
    }
}

File: app/src/main/java/com/hussain/podcastapp/model/Entry.kt
package com.hussain.podcastapp.model

import android.os.Parcel
import android.os.Parcelable
import androidx.annotation.NonNull
import androidx.room.Embedded
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey
import com.google.firebase.database.Exclude
import com.google.gson.annotations.SerializedName
import java.util.*


@Entity(tableName = "entry")
class Entry : Parcelable {
    @NonNull
    @PrimaryKey
    @Embedded
    @SerializedName("id")
    var feedId: FeedID? = null
    @Embedded
    @SerializedName("title")
    @get:Exclude
    @set:Exclude
    var entryTitle: Title? = null
    @SerializedName("im:image")
    var image: List<PodcastImage>? = null
    @Embedded
    @SerializedName("summary")
    var summary: Summary? = null

    constructor()

    @Ignore private constructor(parcel: Parcel) {
        feedId = parcel.readValue(FeedID::class.java.classLoader) as FeedID
        entryTitle = parcel.readValue(Title::class.java.classLoader) as Title
        if (parcel.readByte().toInt() == 0x01) {
            image = ArrayList()
            parcel.readList(image, PodcastImage::class.java.classLoader)
        } else {
            image = null
        }
        summary = parcel.readValue(Summary::class.java.classLoader) as Summary
    }

    override fun describeContents(): Int {
        return 0
    }

    @Exclude
    fun toMap(): Map<String, Any?> {
        val result = HashMap<String, Any?>()
        result["feedId"] = feedId
        result["EntryTitle"] = entryTitle
        result["image"] = image
        result["summary"] = summary
        return result
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeValue(feedId)
        dest.writeValue(entryTitle)
        if (image == null) {
            dest.writeByte(0x00.toByte())
        } else {
            dest.writeByte(0x01.toByte())
            dest.writeList(image)
        }
        dest.writeValue(summary)
    }

    class Title : Parcelable {

        companion object CREATOR : Parcelable.Creator<Title> {
            override fun createFromParcel(parcel: Parcel): Title {
                return Title(parcel)
            }

            override fun newArray(size: Int): Array<Title?> {
                return arrayOfNulls(size)
            }
        }

        @SerializedName("label")
        var labelTitle: String? = ""

        constructor()

        @Ignore
        internal constructor(parcel: Parcel) {
            labelTitle = parcel.readString()
        }

        override fun describeContents(): Int {
            return 0
        }

        override fun writeToParcel(dest: Parcel, flags: Int) {
            dest.writeString(labelTitle)
        }
    }

    class Summary : Parcelable {

        companion object CREATOR : Parcelable.Creator<Summary> {
            override fun createFromParcel(parcel: Parcel): Summary {
                return Summary(parcel)
            }

            override fun newArray(size: Int): Array<Summary?> {
                return arrayOfNulls(size)
            }
        }

        var label: String? = ""

        constructor()

        @Ignore private constructor(parcel: Parcel) {
            label = parcel.readString()
        }

        override fun describeContents(): Int {
            return 0
        }

        override fun writeToParcel(dest: Parcel, flags: Int) {
            dest.writeString(label)
        }
    }

    class FeedID : Parcelable {

        companion object CREATOR : Parcelable.Creator<FeedID> {
            override fun createFromParcel(parcel: Parcel): FeedID {
                return FeedID(parcel)
            }

            override fun newArray(size: Int): Array<FeedID?> {
                return arrayOfNulls(size)
            }
        }

        @Embedded
        @NonNull
        var attributes: Attributes? = null

        constructor()

        @Ignore
        internal constructor(parcel: Parcel) {
            attributes = parcel.readValue(Attributes::class.java.classLoader) as Attributes
        }

        override fun describeContents(): Int {
            return 0
        }

        override fun writeToParcel(dest: Parcel, flags: Int) {
            dest.writeValue(attributes)
        }
    }

    class Attributes : Parcelable {

        companion object CREATOR : Parcelable.Creator<Attributes> {
            override fun createFromParcel(parcel: Parcel): Attributes {
                return Attributes(parcel)
            }

            override fun newArray(size: Int): Array<Attributes?> {
                return arrayOfNulls(size)
            }
        }

        @NonNull
        @SerializedName("im:id")
        var id: String? = ""

        constructor()

        @Ignore private constructor(parcel: Parcel) {
            id = parcel.readString()
        }

        override fun describeContents(): Int {
            return 0
        }

        override fun writeToParcel(dest: Parcel, flags: Int) {
            dest.writeString(id)
        }
    }

    companion object CREATOR : Parcelable.Creator<Entry> {
        override fun createFromParcel(parcel: Parcel): Entry {
            return Entry(parcel)
        }

        override fun newArray(size: Int): Array<Entry?> {
            return arrayOfNulls(size)
        }
    }

}